File: 1503-reducing-dishes/1503-reducing-dishes.py
class Solution:
    def maxSatisfaction(self, sat: List[int]) -> int:
        sat.sort()
        n = len(sat)
        memo = [[None] * (n + 1) for _ in range(n + 1)]
        def dp(idx, count):
            if idx >= len(sat):
                return 0

            if memo[idx][count - 1]:
                return memo[idx][count - 1]

            pick = dp(idx + 1, count + 1) + (count * sat[idx])
            no_pick = dp(idx + 1, count)

            res = max(pick, no_pick)
            memo[idx][count - 1] = res
            return res

        final = dp(0, 1)
        return final

File: 1503-reducing-dishes/1503-reducing-dishes.cpp
class Solution {
public:
    int maxSatisfaction(vector<int>& sat) {
        int n = sat.size();
        int curr = 0;
        int res = 0;
        sort(sat.begin(), sat.end());
        for (int i = n - 1 ; i >= 0 ; i--){
            curr += sat[i];
            if (curr > 0){
                res += curr;
            }
        }
        return res;
    }
};